### PR TITLE
Add unit tests with minimum dependencies

### PR DESCRIPTION
**Pull request by gsheni**
- Add unit tests against minimum dependencies for python 3.7 on PRs (when requirements change) and to run on main.

### CODE REVIEW

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep wheel)"
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep setuptools)"
+          python -m pip install -r evalml/tests/dependency_update_check/minimum_core_requirements.txt
+          python -m pip install -r evalml/tests/dependency_update_check/minimum_requirements.txt
+          python -m pip install -r evalml/tests/dependency_update_check/minimum_test_requirements.txt
+      - if: ${{ matrix.core_dependencies }}
+        name: run tests with minimal dependencies
+        run: |
+          source test_python/bin/activate
+          python -m pytest evalml/ -x -n 2 --doctest-modules --has-minimal-dependencies
```
Comment (ParthivNaresh) [inline .github/workflows/linux_unit_tests_with_minimum_deps.yml:69]: Not a big deal but feel free to add this and the command below to the makefile for easier reference.

**Review by ParthivNaresh — APPROVED** [review]
Looks good to me, nice catch on upgrading numpy to fix the shap issue!

File: evalml/tests/dependency_update_check/minimum_core_requirements.txt [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
+click==7.1.2
 psutil==5.6.3
 requirements-parser==0.2.0
 shap==0.36.0
 texttable==1.6.2
 woodwork==0.0.11
-featuretools==0.20.0
+featuretools==0.21.0
 dask==2.12.0
+distributed==2.12.0
```
Comment (freddyaboulton) [inline evalml/tests/dependency_update_check/minimum_core_requirements.txt:17]: How come we need to add this package? Should be pulled in by featuretools no?

**Review by freddyaboulton — COMMENTED** [review]
@gsheni This looks great! Thanks for splitting up the tests into core and non-core too. I have a couple of questions about why we had to add/bump dependencies before I approve.

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsheni) [reply .github/workflows/linux_unit_tests_with_minimum_deps.yml]: Added the 2 commands to the Makefile

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep setuptools)"
+          python -m pip install -r evalml/tests/dependency_update_check/minimum_core_requirements.txt
+          python -m pip install -r evalml/tests/dependency_update_check/minimum_test_requirements.txt
+      - if: ${{ !matrix.core_dependencies }}
+        name: install evalml with core dependencies, optional requirements, & test dependencies
+        run: |
+          source test_python/bin/activate
+          python -m pip install -e . --no-dependencies
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep wheel)"
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep setuptools)"
```
[the remark this answers is not in the document]
Comment (gsheni) [reply .github/workflows/linux_unit_tests_with_minimum_deps.yml]: Moved test requirements to be installed first

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+          virtualenv test_python -q
+          source test_python/bin/activate
+          python -m pip install --upgrade pip -q
+      - if: ${{ matrix.core_dependencies }}
+        name: install evalml with core dependencies, & test dependencies
+        run: |
+          source test_python/bin/activate
+          python -m pip install -e . --no-dependencies
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep wheel)"
+          python -m pip install "$(cat evalml/tests/dependency_update_check/minimum_test_requirements.txt | grep setuptools)"
```
[the remark this answers is not in the document]
Comment (gsheni) [reply .github/workflows/linux_unit_tests_with_minimum_deps.yml]: Moved test requirements to be installed first, and removed the `python -m` part

File: .github/workflows/linux_unit_tests_with_minimum_deps.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
+          fetch-depth: 2
+      - name: Update apt and install Graphviz
+        run: sudo apt update && sudo apt install -y graphviz
+      - name: Create virtual environment, upgrade pip
+        run: |
+          python -m pip install virtualenv
```
[the remark this answers is not in the document]
Comment (gsheni) [reply .github/workflows/linux_unit_tests_with_minimum_deps.yml]: It's not necessary. Removed it.

File: test-requirements.txt [hunk 6]
```diff
@@ -1,6 +1,8 @@
-pytest==4.4.1
-pytest-xdist==1.26.1
-pytest-cov==2.6.1
+wheel>=0.33.1
```
[the remark this answers is not in the document]
Comment (gsheni) [reply test-requirements.txt:1]: I needed to add `wheel` and `setuptools` to make sure that numpy properly installed. 
In addition, if you don't have `wheel` installed it takes longer to install some packages from PyPI (it does `python setup.py install` for each requirement). So installing wheel first fixes this.

[the remark this answers is not in the document]
Comment (gsheni) [reply test-requirements.txt:1]: I bumped pytest because the old pytest had lower requirements that conflicted with requirements I was bumping.

File: evalml/tests/dependency_update_check/minimum_core_requirements.txt (hunk 2 above)
Comment (gsheni) [reply evalml/tests/dependency_update_check/minimum_core_requirements.txt]: The reason I had to add this is the following pip warning/error (which I saw after installing `minimum_test_requirements.txt`):
```
distributed 2021.5.0 has requirement dask==2021.05.0, but you'll have dask 2.12.0 which is incompatible.
```

File: test-requirements.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply test-requirements.txt:1]: Got it!

**Review by freddyaboulton — APPROVED** [review]
Thank you @gsheni !!
